baldrick/plugins/github_pull_requests.py

```diff
@@ -92,38 +95,49 @@ def is_previous_comment(message):
-        if f_status is not None:
-            set_status = True
-            status = status and f_status
-        comments += f_comments
-
-    if comments:
-        message = current_app.pull_request_prolog.format(pr_handler=pr_handler, repo_handler=repo_handler)
-        message += ''.join(comments) + current_app.pull_request_epilog
-        comment_url = pr_handler.submit_comment(message, comment_id=comment_id,
-                                                return_url=True)
-    else:
-        all_passed_message = pr_config.get("all_passed_message", '')
-        all_passed_message = all_passed_message.format(pr_handler=pr_handler, repo_handler=repo_handler)
-        if all_passed_message:
-            pr_handler.submit_comment(all_passed_message, comment_id=comment_id)
+        result = function(pr_handler, repo_handler)
+        results.update(result)
 
-        comment_url = None
-        message = ''
+    failures = [message for message, status in results.values() if not status]
+
+    if post_comment:
+
+        # Post all failures in a comment, and have a single status check
+
+        if failures:
+
+            pull_request_prologue = get_config_with_app_defaults(pr_handler, 'pull_request_prologue', default='')
+            pull_request_epilogue = get_config_with_app_defaults(pr_handler, 'pull_request_epilogue', default='')
+
+            message = pull_request_prologue.format(pr_handler=pr_handler, repo_handler=repo_handler)
+            message += ''.join(failures) + pull_request_epilogue
+            comment_url = pr_handler.submit_comment(message, comment_id=comment_id,
+                                                    return_url=True)
 
-    if set_status:
-        if status:
-            pr_handler.set_status('success',
-                                  pr_config.get("pr_passed_status", "Passed all checks"),
-                                  current_app.bot_username, target_url=comment_url)
-        else:
             pr_handler.set_status('failure',
                                   pr_config.get("pr_failed_status", "Failed some checks"),
                                   current_app.bot_username, target_url=comment_url)
 
-    return message
+        else:
+
+            all_passed_message = pr_config.get("all_passed_message", '')
+            all_passed_message = all_passed_message.format(pr_handler=pr_handler, repo_handler=repo_handler)
+            if all_passed_message:
+                pr_handler.submit_comment(all_passed_message, comment_id=comment_id)
+
+            pr_handler.set_status('success',
+                                  pr_config.get("pr_passed_status", "Passed all checks"),
+                                  current_app.bot_username)
+
+    else:
+
+        # Post each failure as a status
+
+        for context, (message, status) in sorted(results.items()):
+
+            pr_handler.set_status('success' if status else 'failure', message,
```

Comment:
rather than having the plugins pass back a bool should we not just pass though a suitable string? I could see a plugin wanting to set `pending` on one hook and `success` on another?

Reply:
Yeah good point